corgidrp/calibrate_kgain.py

```diff
@@ -0,0 +1,924 @@
+    return popt, lambda params: (sse, gauss(xdata, *params))
+    
+def Single_peakfit(xdata, ydata):
+    """
+    Fit Gauss function to x, y data. Returns mean and sigma. Only sigma 
+    is used in main code call.
+    
+    Args:
+      xdata (np.array): x data
+      ydata (np.array): y data
+    
+    Returns:
+      float: sigma
+    """
+    astart = np.max(ydata)
+    mustart = xdata[np.argmax(ydata)]
+    sigmastart = 10
+    sgaussinp = [astart, mustart, sigmastart]
+    
+    estimates, model = sgaussfit(xdata, ydata, sgaussinp)
+    a1, mu1, sigma = estimates
+
+    return sigma
+    
+def histc_roi(frame,offset_rowroi,offset_colroi,rn_bins):
+    """
+    Histogram of pixel values of frame within the ROI and bins defined in 
+    rn_bins. Returns the counts in each bin.
+    
+    Args:
+      frame (np.array): frame
+      offset_rowroi (int): row of region of interest
+      offset_colroi (int): column of region of interest
+      rn_bins (int): histogram bins
+    
+    Returns:
+      np.array: counts in each bin
+    """
+    selected_area = frame[offset_rowroi,offset_colroi]
+    data_reshaped = selected_area.ravel()
+    counts, _ = np.histogram(data_reshaped, bins=rn_bins)
+      
+    return counts
+    
+def calculate_mode(arr):
+    """
+    calculates histogram of an array
+    
+    Args:
+      arr (np.array): input array
+    
+    Returns:
+      np.array: bin center values
+      np.array: bin center counts
+    """
+    counts, bin_edges = np.histogram(arr)
+    # Calculate bin centers (values)
+    values = (bin_edges[:-1] + bin_edges[1:]) / 2
+    max_count_index = np.argmax(counts)
+    return values[max_count_index], counts[max_count_index]
+
+def sigma_clip(data, sigma=2.5, max_iters=6):
+    """
+    Perform sigma-clipping on the data.
+      
+    Args:
+       data (np.array): The input data to be sigma-clipped.
+       sigma (float): The number of standard deviations to use for clipping.
+       max_iters (int): The maximum number of iterations to perform.
+    
+    Returns:
+      np.ndarray: The sigma-clipped data.
+      np.ndarray: A boolean mask where True indicates a clipped value.
+    """
+    data = np.asarray(data)
+    clipped_data = data.copy()
+      
+    for i in range(max_iters):
+        mean = np.mean(clipped_data)
+        std = np.std(clipped_data)
+        mask = np.abs(clipped_data - mean) > sigma * std
+        if not np.any(mask):
+            break
+        clipped_data = clipped_data[~mask]
+        
+    return clipped_data, mask
+    
+######################### start of main code #############################
+
+def calibrate_kgain(dataset_kgain, actual_gain, actual_gain_mean_frame,
+                    n_cal=10, n_mean=30, min_val=800, max_val=3000, binwidth=68,
+                    make_plot=True,plot_outdir='figures', show_plot=False,
+                    logspace_start=-1, logspace_stop=4, logspace_num=200, verbose=False):
+    """
```

Comment:
@maxwellmb: Are you suggesting to use the actual gain values in CMDGAIN? That could be an issue. First, there are checks for it to be 1 in Kgain, and measured values need not be exactly 1.0. More importantly, in non-linearity the values of CMDGAIN are used to split the dataset. This only works if there's a set with the same value (commanded). The actual values may be different by fractional values and the split would not work as expected.

Can we not just create a new Keyowrd, as I had to to with the mean frame, and update the Issue https://github.com/roman-corgi/corgidrp/issues/136 accordingly? And add this keyword, together with the one of the mean frame to the list of desired keywords? They can be renamed at any time and scripts be updated.

Reply:
Sorry if that was confusing.  We should definitely not change the CMDGAIN, but you should have your code use CMDGAIN for now since we don't have any actual gain keywords or infrastructure implemented. I'd prefer this to adding an extra keyword for now. However if you could comment on the issues where this is relevant so that we can keep track of places where we need to change it.

Reply:
Yes, I think anywhere arrays are sorted according to EM gain, CMDGAIN is all that is needed.  Anywhere an actual EM gain value is utilized (divided by, for example), a more accurate EM gain if available should be used.  And the good thing is that a commanded gain of 1.0 should be physically an applied gain of 1 (i.e., no gain applied), by definition, regardless of the parametrization of the EM gain curve, and that's because no voltage was applied to the HV electrode in that case.

Reply:
Agreement (Max) is to follow what Kevin says/did: "eah, the CMDGAIN is supposed to be the actual applied gain, assuming a good fit to the parametrization function.  In the future, we may implement a methodology of reading off the EM gain from a given frame and assign some new header like ESTGAIN or EMGAIN_M for that read-off value.  But since that is still a little up in the air, and since CMDGAIN in practice should be close enough to the applied gain, he wanted us to just use CMDGAIN for now.
If you really wanted to, you could do something like I did in my calibrate_darks_lsq() function with a try and except statement so that even if EMGAIN_M is never implemented, the code doesn't fail:
```
try: # if EM gain measured directly from frame TODO change hdr name if necessary
     EMgain_arr = np.append(EMgain_arr, datasets[i].frames[0].ext_hdr['EMGAIN_M'])
except: # use commanded gain otherwise TODO change hdr name if necessary
     EMgain_arr = np.append(EMgain_arr, datasets[i].frames[0].ext_hdr['CMDGAIN']):"
```

Reply:
Done.